layer2-state-provider/history_commitment_provider.go

```diff
@@ -441,7 +441,7 @@ func (p *HistoryCommitmentProvider) OneStepProofData(
 
 	// Compute the exact start point of where we need to execute
 	// the machine from the inputs, and figure out, in what increments, we need to do so.
-	machineIndex, err := p.computeMachineStartIndex(startHeights)
+	machineIndex, err := p.computeMachineStartIndex(startHeights, Height(assertionMetadata.FromState.PosInBatch))
```

Comment:
I'm not convinced this is right. Why is the PosInBatch relevant to the opcode index we need?

Reply:
Yeah. I 'm thinking this isn't right either. I think the problem is that while we have some unit tests which exercise that the math is right. We aren't actually exercising that the API makes sense when dealing with a one-step proof of some opcode somewhere in in the small-step range. So, we don't see how obviously broken this this code is.

I'm going to see how hard it is to add such an end-to-end test now.

Reply:
Yup. I finally realized that in the case of a "OneStepProof" the offset into the challenge-level set of opcodes we are working on, IS the `upToHeight` argument.